src/Tooltip/Tooltip.js

```diff
@@ -0,0 +1,106 @@
+// @flow
+
+import * as React from 'react';
+import { Dimensions, View } from 'react-native';
+import type { ViewLayoutEvent } from 'react-native/Libraries/Components/View/ViewPropTypes';
+import TooltipBubble from './TooltipBubble';
+import StyleSheet from '../PlatformStyleSheet';
+
+type Props = {|
+  +isActive?: boolean,
+  +arrowHorizontalPosition?: 'flex-start' | 'center' | 'flex-end',
+  +children: React.Node,
+  +text: string,
+  +verticalShift?: number,
+|};
+
+type State = {|
+  elementY: number,
+  childrenHeight: number,
+|};
+
+export default class Tooltip extends React.Component<Props, State> {
+  childrenRef: ?React$ElementRef<Object>;
+
+  static defaultProps = {
+    isActive: false,
+    arrowHorizontalPosition: 'center',
+    verticalShift: 0,
+  };
+
+  state = {
+    elementY: 0,
+    childrenHeight: 0,
+  };
+
+  componentDidUpdate() {
+    this.measureChildrenHeight();
+  }
+
+  measureChildrenHeight = () => {
+    this.childrenRef &&
+      this.childrenRef.measure((a, b, width, height) => {
+        const roundedHeight = Math.round(height);
+        const { childrenHeight } = this.state;
+        roundedHeight !== childrenHeight &&
+          this.setState({ childrenHeight: roundedHeight });
+      });
+  };
+
+  onLayout = (event: ViewLayoutEvent) => {
+    const { layout } = event.nativeEvent;
+    this.setState({ elementY: layout.y });
+  };
+
+  getChildrenRef = (ref: ?React$ElementRef<Object>) => {
+    this.childrenRef = ref;
+  };
+
+  render() {
+    const screenHeight = Dimensions.get('window').height;
+    const {
+      arrowHorizontalPosition,
+      children,
+      isActive,
+      text,
+      verticalShift,
+    } = this.props;
+    const { elementY, childrenHeight } = this.state;
+    const isArrowDown = screenHeight / 2 > elementY;
+    const computedStyles = getComputedStyles(isArrowDown);
+
+    return (
+      <View onLayout={this.onLayout} style={styles.tooltipComponent}>
+        <View ref={this.getChildrenRef} renderToHardwareTextureAndroid>
+          {children}
+        </View>
+        {isActive && childrenHeight !== 0 && (
+          <TooltipBubble
+            text={text}
+            arrowVerticalPosition={isArrowDown ? 'flex-start' : 'flex-end'}
+            arrowHorizontalPosition={arrowHorizontalPosition}
+            style={computedStyles.tooltipBubble}
+            verticalShift={verticalShift + childrenHeight}
+          />
+        )}
+      </View>
+    );
+  }
+}
+
+const styles = StyleSheet.create({
+  tooltipComponent: {
+    position: 'relative',
+    zIndex: 100,
+  },
+});
+
+const getComputedStyles = isArrowDown =>
```

Comment:
@tbergq I am still not sure what the best way is to avoid recreating styles on every render? Is that enough because `StyleSheet.create` will somehow cache the styles?

Reply:
With this approach, the styles will be cached inside of the render method of the Tooltip component. The cached style is passed down, and if a child component re-renders, it will use the cached value. If tooltip re-renders, it will create a new cached value. 

I would rather split this into something like

```
const styles = StyleSheet.create({
  tooltipComponent: {
    position: 'relative',
    zIndex: 100,
  },
tooltipBubble: {
      position: 'absolute',
      zIndex: 101,
      opacity: 1,
    },
bubbleArrowDown: {top: 0},
bubbleArrowUp: {bottom: 0},
});
```

then using `style={[styles.tooltipBubble, isArrowDown ? styles.bubbleArrowDown : styles.bubbleArrowUp]}`

Reply:
@tbergq That make sense, but what if it's more complicated and I need to pass the argument to the style and decide which style key should be changed?

Here is the example:
```jsx
<View
  style={[
    styles.invisibleContainer,
    {
      [arrowVerticalPosition === 'flex-start'
        ? 'marginTop'
        : 'marginBottom']: verticalShift,
    },
  ]}
>
```
It's messy to have have those inline styles, so it make sense to separate style logic outside the JSX. 


What do you think about following solution?
```jsx
<View
  style={[
    styles.invisibleContainer,
    arrowVerticalPosition === 'flex-start'
      ? getComputedStyles(verticalShift).invisibleContainerTopShift
      : getComputedStyles(verticalShift).invisibleContainerBottomShift,
    style,
  ]}
>
```
```js
const getComputedStyles = verticalShift => StyleSheet.create({
  invisibleContainerTopShift: {
    marginTop: verticalShift,
  },
  invisibleContainerBottomShift: {
    marginBottom: verticalShift,
  },
});

```

Btw I'm not sure if it's necessary to use `StyleSheet.create` when I calculate styles depending on some parameters.

Reply:
Well, you could still put the logic of deciding on the correct style in the `render` method before the `return` statement. That means you can create all your styles in advance, have the logic in the render method to decide which one to use, and then if some are too dynamic to have been created in advance, you just add a dynamic property. WDYT?

Reply:
That sounds better, I will do it that way.

Reply:
On a general basis, if there are few options, use cached styles. If it is many I am ok with using inline styles. But it should be kept to a minimum 😊